### PR TITLE
Use poetry for bootstrap

### LINKED ISSUE #468: Use poetry for bootstrap dependency mgmt
Once #467 lands, we can migrate the bootstrap dependency management from `pip-compile` to `poetry`.

### PR DESCRIPTION
**Pull request by eloquence**
Resolves #468

This replaces our use of `pip-compile` with `poetry` to manage the dependencies for building other wheels, and updates the instructions accordingly.

To do so, it freezes all bootstrap dependencies at the current state, except for the ones we no longer need. 

## Context

Our bootstrap is quite fragile due to being so dated, and changes to individual dependencies can either cause breakage, changes to existing wheels, or both. Most notably, we cannot update to the latest `Cython` until we have removed our use of PyYAML (https://github.com/freedomofpress/securedrop-client/issues/1681), which is underway. Instead of opportunistic piecemeal updates, #464 proposes a full update of all bootstrap requirements, which I would suggest we do after we've removed PyYAML so we don't have to worry about that piece.

## Testing

To test this PR, try rebuilding the bootstrap by following the instructions in the README, with a test dependency such as `cowsay`, while on this branch:

```
# Ensure you are running in a cleanly boostrapped virtual environment
rm -rf .venv
make install-deps
source .venv/bin/activate
# Update the workstation-bootstrap/pyproject.toml file with a new dependency
# For example, add `cowsay = "6.0.0"` to `pyproject.toml`
# Then update the lockfile:
poetry -C workstation-bootstrap/ lock
# Now we are ready to build updated wheels:
./scripts/build-sync-wheels --project workstation-bootstrap --pkg-dir ./workstation-bootstrap
# Once the new wheels are ready, we recreate our sha256sums:
./scripts/sync-sha256sums ./workstation-bootstrap
# Sign the list of sha256sums
gpg --armor --output workstation-bootstrap/sha256sums.txt.asc --detach-sig  workstation-bootstrap/sha256sums.txt
# We can even verify if we want
./scripts/verify-sha256sum-signature ./workstation-bootstrap/
# Update the build-requirements.txt file
./scripts/update-requirements --pkg-dir ./workstation-bootstrap/ --project workstation-bootstrap
```

Check if you encounter errors or unexpected results (`git diff`) along the way.

### CODE REVIEW

File: workstation-bootstrap/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+[tool.poetry.dependencies]
+python = "^3.9"
+build = "0.3.0"
+wheel = "0.36.2"
+pip = "21.3.1"
+setuptools-scm = "5.0.2"
+pytest-runner = "5.3.0"
+cython = "0.29.33"
+flit-core = "2.3.0"
+# Sub-dependencies which we're explicitly freezing for now,
```
Comment (legoktm) [inline workstation-bootstrap/pyproject.toml:18]: Now that you've pinned these at the correct versions in the lock file, you can remove them and run `poetry lock --no-update` and then it'll keep them at the correct versions without needing them to be explicitly specified.

File: README.md [hunk 2]
```diff
@@ -70,25 +70,23 @@ We use [build](https://pypa-build.readthedocs.io/en/latest/) toolchain to build
...
-# Then update the requirements.in file as required
-pip-compile --allow-unsafe --generate-hashes \
-    --output-file=workstation-bootstrap/requirements.txt workstation-bootstrap/requirements.in
-# Now we are ready for bootstrapping
-./scripts/build-sync-wheels --project workstation-bootstrap --pkg-dir ./workstation-bootstrap --requirements .
-# Here we have the new wheels ready
-# Now let us recreate our new sha256sums for bootstrapping
+# Update the workstation-bootstrap/pyproject.toml file as required.
+# Then update the lockfile:
+poetry -C workstation-bootstrap/ lock
```
Comment (legoktm) [inline README.md:79]: I think the workflow should be either: run `poetry update <name>` OR update pyproject.toml + `poetry lock --no-update`

**Review by legoktm — COMMENTED** [review]
Nice :) walked through the test plan and it all seemed to work!

File: workstation-bootstrap/pyproject.toml (hunk 1 above)
Comment (eloquence) [reply workstation-bootstrap/pyproject.toml]: Good point, done!

File: README.md (hunk 2 above)
Comment (eloquence) [reply README.md]: Makes sense to me, fleshed out README examples a bit.

**Review by legoktm — APPROVED** [review]
Awesome!!

**Pull request merged**